cherrypy/test/test_client_cert.py

```diff
@@ -0,0 +1,140 @@
+
+def subdict(d, key_predicate):
+    return {k: v for k, v in d.items() if key_predicate(k)}
+
+
+class ClientCertTests(object):
+    """Client certificate validation tests."""
+    scheme = 'https'
+    script_name = '/wsgi_env'
+
+    server_should_reject = False
+    server_host = 'localhost'
+    client_host = 'localhost'
+    ca_chain = CA_CERT
+    client_cert = CLIENT_CERT
+    verify_mode = None  # None, 'none', 'optional', 'required'
+
+    @classmethod
+    def setup_server(cls):
+        """Prepare server for test."""
+        cherrypy.config.update({
+            'checker.on': False,
+            'engine.autoreload_on': False,
+            'server.socket_host': cls.server_host,
+            'server.ssl_module': 'builtin',
+            'server.ssl_private_key': SERVER_KEY,
+            'server.ssl_certificate': SERVER_CERT,
+            'server.ssl_certificate_chain': cls.ca_chain,
+            'server.ssl_verify_mode': cls.verify_mode,
+        })
+
+        class Root:
+            """Test application."""
+            @cherrypy.expose
+            @cherrypy.tools.json_out()
+            def wsgi_env(self):
+                """Return client certificate WSGI environment variables."""
+                return subdict(
+                    cherrypy.request.wsgi_environ,
+                    lambda k: k.startswith('SSL_CLIENT_'))
+
+        cherrypy.tree.mount(Root())
+
+    def test_connect(self):
+        """If server_should_reject is false, verify the index can be fetched.
+        Otherwise, verify that a URLError is raised when connecting.
+        """
+        context = ssl.create_default_context(cafile=self.ca_chain)
+        context.load_cert_chain(self.client_cert, keyfile=CLIENT_KEY)
+        context.check_hostname = False
+
+        if self.server_should_reject:
+            with pytest.raises(urllib.error.URLError):
+                urllib.request.urlopen(self.base(), context=context)
+        else:
+            self.assert_wsgi_env(context)
+
+    def assert_wsgi_env(self, ssl_context):
+        """Verify the response contains WSGI environment variables from our client certificate."""
+        if self.verify_mode in (None, 'none'):
+            return
+
+        body = urllib.request.urlopen(self.base(), context=ssl_context).read()
+        wsgi_env = json.loads(body.decode())
+        assert wsgi_env['SSL_CLIENT_S_DN_C'] == 'US'
+        assert wsgi_env['SSL_CLIENT_S_DN_CN'] is not None
+        assert wsgi_env['SSL_CLIENT_S_DN_O'] == 'CherryPy'
+        assert wsgi_env['SSL_CLIENT_S_DN_ST'] == 'XX'
+        assert wsgi_env['SSL_CLIENT_I_DN_C'] == 'US'
+        assert wsgi_env['SSL_CLIENT_I_DN_L'] == 'XXX'
+        assert wsgi_env['SSL_CLIENT_I_DN_O'] == 'CherryPy'
+        assert wsgi_env['SSL_CLIENT_I_DN_ST'] == 'XX'
+
+
+BAD_CLIENT_CERTS = [CLIENT_WRONG_CA]
+GOOD_CLIENT_CERTS = [CLIENT_CERT, CLIENT_IP_CERT, CLIENT_WILD_CERT]
+SERVER_HOSTS = ['localhost', '127.0.0.1', '0.0.0.0']
+CLIENT_HOSTS = ['localhost']
+
+TESTS = [
```

Comment:
Do you think you could rewrite this using `@pytest.parametrize`?

Reply:
I'll give it a shot.

Reply:
I tried changing ClientCertTests to inherit from CPWebCase, and added a parameterized test function to it, but pytest parameterize does not work on unittest.TestCase test methods.  My other ideas to do this are bad, requiring code duplication from CPWebCase.  Do you have any suggestions?